### PR TITLE
DBZ-6885 Support for GKE workload identities

### PR DESCRIPTION
**Pull request by laughingman7743**
### Which use case/requirement will be addressed by the proposed feature?

We are building KafkaConnect on GKE and GKE has a mechanism to authenticate without issuing a ServiceAccount JSON key.
https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity
Since issuing JSON keys is a security risk, we would like to support GKE workload identities with Spanner connectors.

### Implementation ideas (optional)

https://cloud.google.com/docs/authentication
If no JSON key is explicitly set, the Application Default Credentials mechanism can be used for proper authentication processing.

https://issues.redhat.com/browse/DBZ-6885

### CODE REVIEW

File: src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java [hunk 1]
```diff
@@ -181,7 +181,7 @@ public abstract class BaseSpannerConnectorConfig extends CommonConnectorConfig {
             .withType(Type.STRING)
             .withGroup(Field.createGroupEntry(Field.Group.CONNECTOR, 1))
             .withWidth(Width.MEDIUM)
-            .withImportance(Importance.HIGH)
+            .withImportance(Importance.LOW)
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java:184]: IMHO if reduction of importance is done then it should be `MEDIUM`. Same for the one below.

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java [hunk 2]
```diff
@@ -56,10 +56,7 @@ public ConnectionValidator validate() {
         if (!FieldValidator.isSpecified(googleCredentials) && !FieldValidator.isSpecified(credentialPath) && !FieldValidator.isSpecified(credentialJson)) {
             String message = String.format(PLEASE_SPECIFY_CONFIGURATION_PROPERTY_MSG, SPANNER_CREDENTIALS_PATH.name(),
                     SPANNER_CREDENTIALS_JSON.name(), GOOGLE_APPLICATION_CREDENTIALS_ENV_VAR);
-            LOGGER.error(message);
-            context.error(message, SPANNER_CREDENTIALS_PATH, SPANNER_CREDENTIALS_JSON);
-            result = false;
-            return this;
+            LOGGER.warn(message, SPANNER_CREDENTIALS_PATH, SPANNER_CREDENTIALS_JSON);
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:59]: It should not be warning, just `INFO`.
Also the validation should be kept there but modified. If any of the settings is set but not all then there should be validation error. If none is set then the App Default Credentials will be used.

File: src/main/java/io/debezium/connector/spanner/db/DatabaseClientFactory.java [hunk 3]
```diff
@@ -96,6 +97,15 @@ else if (credentialsPath != null) {
                 LOGGER.error(e.getMessage(), e);
             }
         }
+        else {
```
Comment (jpechane) [inline src/main/java/io/debezium/connector/spanner/db/DatabaseClientFactory.java:100]: This method is for testing. Do you have any test for that? If not then the code should not be added.

**Review by jpechane — CHANGES_REQUESTED** [review]
@laughingman7743 Thanks for the PR! I left couple of comments.

@nancyxu123 Are you fine with the enahncement in general?

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java (hunk 2 above)
Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java]: Fixed: https://github.com/debezium/debezium-connector-spanner/pull/65/commits/678725a918c7c947a8fba9d223795b82c2072f69

File: src/main/java/io/debezium/connector/spanner/db/DatabaseClientFactory.java (hunk 3 above)
Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/db/DatabaseClientFactory.java:100]: https://github.com/debezium/debezium-connector-spanner/pull/65/commits/32a1f61b6c65971377dd22f4a7d124915b2bb1a2
I added a test case when credentialJson and credentialFile are null. I am not sure if it is a good idea to add such a test case. Please let me know if there is an appropriate test case.

File: src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java (hunk 1 above)
Comment (nancyxu123) [reply src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java]: Why would reduction in importance be needed here?

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java (hunk 2 above)
Comment (nancyxu123) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java]: I'm wondering if we should add validation here that if there is no credentials JSON / path specified, there should be application default credentials specified. If not, we should throw an error.

Then, we can probably add a unittest here?

https://github.com/debezium/debezium-connector-spanner/blob/9306f1349643cf1d1c56ed1c335615a5f916cb2e/src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java#L22

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java [hunk 4]
```diff
@@ -56,10 +56,7 @@ public ConnectionValidator validate() {
         if (!FieldValidator.isSpecified(googleCredentials) && !FieldValidator.isSpecified(credentialPath) && !FieldValidator.isSpecified(credentialJson)) {
             String message = String.format(PLEASE_SPECIFY_CONFIGURATION_PROPERTY_MSG, SPANNER_CREDENTIALS_PATH.name(),
                     SPANNER_CREDENTIALS_JSON.name(), GOOGLE_APPLICATION_CREDENTIALS_ENV_VAR);
-            LOGGER.error(message);
-            context.error(message, SPANNER_CREDENTIALS_PATH, SPANNER_CREDENTIALS_JSON);
-            result = false;
```
Comment (nancyxu123) [inline src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: Why did you remove false here?

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java [hunk 5]
```diff
@@ -22,7 +22,7 @@ public class ConnectionValidator implements ConfigurationValidator.Validator {
 
     private static final String GOOGLE_APPLICATION_CREDENTIALS_ENV_VAR = "GOOGLE_APPLICATION_CREDENTIALS";
```
Comment (nancyxu123) [inline src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:23]: Would the environment variable here be enough to detect application deffault credentials?

File: src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java (hunk 1 above)
Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/BaseSpannerConnectorConfig.java]: These JSON key options are not required when using workload identity authentication. The importance should be changed.

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java (hunk 2 above)
Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java]: https://github.com/debezium/debezium-connector-spanner/blob/9306f1349643cf1d1c56ed1c335615a5f916cb2e/src/test/java/io/debezium/connector/spanner/config/validation/ConnectionValidatorTest.java#L38
This test case is Disabled. Since the validation has been changed to just check the key and output the INFO log, I don't see the need to change the current test.

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java (hunk 4 above)
Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: When using workload identity authentication, it is not necessary to set an authentication key. Such authentication methods must be supported.
https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#authenticating_to

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java (hunk 5 above)
Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:23]: Depending on the authentication method, this environment variable may not be set.
https://cloud.google.com/docs/authentication#auth-decision-tree

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java (hunk 4 above)
Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: https://cloud.google.com/docs/authentication/provide-credentials-adc
> Note: Service account keys are a security risk if not managed correctly. You should[ choose a more secure alternative to service account keys](https://cloud.google.com/docs/authentication#auth-decision-tree) whenever possible. If you must authenticate with a service account key, you are responsible for the security of the private key and for other management operations such as [key rotation](https://cloud.google.com/iam/docs/key-rotation). For more information, see [best practices for managing service account keys](https://cloud.google.com/iam/docs/best-practices-for-managing-service-account-keys).

**Comment by laughingman7743** [comment]
We are already running this patched Spanner connector on our GKE cluster and it is working fine without JSON key configuration.

File: src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java (hunk 4 above)
Comment (nancyxu123) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: Sorry, I think it's best here then to move the validation to check for the application default credentials here if the JSON key doesn't exist. Then, it's best to validate the new logic with a new unittest. Thanks.

Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: OK, I understand what you want to do. If the key is not set in the validation, how about calling `ServiceAccountCredentials.getApplicationDefault()` and checking the application default credentials setting?
If `IOException` is raised, I think we can make it a validation error.

Comment (jpechane) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: I think one of my proposals was lost above. How about chaning the validation logic this way

- if none of Google credentials, credential path and credential JSON is set the log INFO message tha application default credentials will be used
- if at least one but not all are set then emit the current validation error

@nancyxu123 WDYT?

Comment (nancyxu123) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: Hi Jiri! I'm a bit confused here, could I ask for a quick clarification?

My understanding was that at least one of Google credentials, credential path, credentials JSON, or application default credentials should be set. 

If none of them are set, we should emit an error. Otherwise, we should use one of them. Is that correct? @laughingman7743

Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: Yes, if the JSON key is not set, check the ApplicationDefaultCredentials and make an error. Is the following implementation acceptable?
https://github.com/debezium/debezium-connector-spanner/pull/65/commits/99e9087a34455baee7666fb1ae92a24118ab6294

Comment (jpechane) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: @nancyxu123 I am sorry you are right, I just misread the condition :-(.

Comment (nancyxu123) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: This looks great, thank you so much! @laughingman7743 @jpechane

Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: https://github.com/debezium/debezium-connector-spanner/pull/65/commits/5773cd89ab6788eab6c1be150d8cd79de2d3e098
I have added some test cases. Please check them.

For some reason the old versions of mockito-core and mockito-inline were used, so I explicitly added them to the dependencies. (Older versions cannot use mockStatic.)
https://github.com/debezium/debezium-connector-spanner/pull/65/commits/5773cd89ab6788eab6c1be150d8cd79de2d3e098#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R191-R200

Comment (laughingman7743) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: I think we may as well add a ConnectionValidator test for when the environment variable `GOOGLE_APPLICATION_CREDENTIALS` is present.
However, if we want to easily change the environment variables with tests, it should be better to add libraries such as SystemRule, SystemLambda, SystemStubs or JUnitPioneer.
https://stackoverflow.com/questions/8168884/how-to-test-code-dependent-on-environment-variables-using-junit
https://github.com/stefanbirkner/system-rules
https://github.com/stefanbirkner/system-lambda/
https://github.com/webcompere/system-stubs
https://github.com/junit-pioneer/junit-pioneer
If you are willing to add the libraries I would also add the environment variable tests.

Comment (jpechane) [reply src/main/java/io/debezium/connector/spanner/config/validation/ConnectionValidator.java:61]: @laughingman7743 I don't think this is necessary to test it that much.

**Comment by jpechane** [comment]
@laughingman7743 Applied, thanks!
